### PR TITLE
Split Settings::Campaign from SiteConfig

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Refactor

## Description

This PR extracts campaign settings from `SiteConfig` into `Settings::Campaign` and changes all related code.

## Related Tickets & Documents

https://github.com/forem/rfcs/pull/126

## QA Instructions, Screenshots, Recordings

* run `bin/setup` (or `rails data_updates:run`) and check if the data has been correctly migrated
* Go to `/admin/config` and make sure the campaign settings behave as expected.

### UI accessibility concerns?

None, as no user-facing behavior was changed, only backend implementation details.

## Added tests?

- [x]  Yes, added/updated as appropriate

## [Forem core team only] How will this change be communicated?

- [X] This change does not need to be communicated, and this is why not: this is purely a backend refactoring without user-facing changes

## [optional] Are there any post deployment tasks we need to perform?

- [ ] Keep an eye on our error rate to ensure this works as expected

### CODE REVIEW

**Comment by citizen428** [comment]
Wow, this took quite a bit of effort to beat into shape after the other branch got merged, good reminder to next time do something else instead of branching off a big branch 😅

File: app/lib/constants/settings.rb
Comment (rhymes) [inline app/lib/constants/settings.rb:13]: the ultimate URL regexp :D

File: app/controllers/admin/settings/campaigns_controller.rb
Comment (rhymes) [inline app/controllers/admin/settings/campaigns_controller.rb:9]: I wonder if `Site configuration` is the correct phrasing, but we might want to revisit this once we landed on proper naming in the admin restructure. Definitely premature now

File: app/models/campaign.rb
Comment (rhymes) [inline app/models/campaign.rb:22]: 🔥

File: app/models/settings/campaign.rb
Comment (rhymes) [inline app/models/settings/campaign.rb:9]: Isn't this the same regexp in `Constants::Settings` ?

File: app/services/campaigns/settings_upsert.rb
Comment (rhymes) [inline app/services/campaigns/settings_upsert.rb:27]: 👍

File: app/models/site_config.rb
Comment (rhymes) [inline app/models/site_config.rb:49]: Removing the fields will make the migration script fail

**Review by rhymes — COMMENTED** [review]
This is moving along nicely! Left a few notes

File: app/lib/constants/settings.rb
Comment (citizen428) [reply app/lib/constants/settings.rb:13]: It just moved from `SiteConfig` because I need to reference it in several places now.

File: app/controllers/admin/settings/campaigns_controller.rb
Comment (citizen428) [reply app/controllers/admin/settings/campaigns_controller.rb:9]: I wonder that for a while and figured I'll just switch it to "Settings were successfully updated" at the end. But then we probably should also change the menu entry which currently says "Config".

File: app/models/site_config.rb
Comment (citizen428) [reply app/models/site_config.rb:49]: Right. Second time in a row I do this even though I **know** it won't work. 

image

File: app/models/settings/campaign.rb
Comment (citizen428) [reply app/models/settings/campaign.rb:9]: I deliberated for a while where to put this, in the end I settled on a new `UrlValidator` because it makes the most sense and we did custom validators for things like the emoji regex too.

**Review by rhymes — COMMENTED** [review]
Sorry for not catching this sooner. A note, we already have a URL validator:

https://github.com/forem/forem/blob/25bf0e395025f23c7d3389bf177973e44ea6dfb0/Gemfile#L115

which we use in `Article`, `Organization`, `Event`, `Sponsorship` and `GithubRepo`.

Not sure if we should have two of them...

**Comment by citizen428** [comment]
Right. I guess the reason we weren't using this before is that `SiteConfig` validations originally were done manually in a controller so when we moved them into the model things just got copied over. I have no problems switching from my custom solution to this, let me try if it works.

**Comment by citizen428** [comment]
@rhymes Replaced the URL matcher and updated the assertions to match the different error message, this should be ready now.

**Review by rhymes — COMMENTED** [review]
I tried to update the call to action and the expiry time but it doesn't work.

I found this notice in the logs:

> web       | Unpermitted parameters: :hero_html_variant_name, :articles_require_approval, :sidebar_enabled, :sidebar_image, :url, :featured_tags, :call_to_action, :articles_expiry_time

**Comment by citizen428** [comment]
> We should probably add a test or two :D

This is actually covered by tests (the admin config request specs). Unless we add a systems test (which I'm not keen on) this is the one case that's miss-able because the space got updated correctly.

**Comment by citizen428** [comment]
@rhymes Update the view now, but directly in the Github web UI, hope I didn't mess things up.

**Comment by rhymes** [comment]
> > We should probably add a test or two :D
> 
> This is actually covered by tests (the admin config request specs). Unless we add a systems test (which I'm not keen on) this is the one case that's miss-able because the spec got updated correctly.

Ah ok, probably we should add e2e tests. If we hadn't caught it sooner it might have been deployed to all the Forem fleet.
We have one test for the authentication section, but not for the others: https://github.com/forem/forem/blob/master/cypress/integration/adminFlows/config/authenticationSection.spec.js

**Comment by citizen428** [comment]
> Ah ok, probably we should add e2e tests. If we hadn't caught it sooner it might have been deployed to all the Forem fleet.

Hmm, maybe. I don't feel like these tests add too much value long term and I "lost" a lot of time during the first PR waiting for it to get merged and having to repeatedly rebase and merge it, so I'd also just like to get this over with slowly. If we really had deployed this it would "only" have affected setting a handful of config values for admin and it would have been fixed within a couple of minutes.

I'm people absolutely insist on e2e tests I'll add them of course but I'm not convinced by their value.

**Comment by citizen428** [comment]
@rhymes I added an e2e test, verifying that URL validation for the campaign image works, this one seemed at least somewhat useful (though we test this on the Rails side too IIRC).

**Review by rhymes — APPROVED** [review]
LGTM!

File: app/controllers/admin/settings/campaigns_controller.rb
Comment (jgaskins) [inline app/controllers/admin/settings/campaigns_controller.rb:11]: [nonblocking] I've always done `errors.full_messages.to_sentence` so it took me a bit of reading through this PR to realize that this is returning an array of error messages instead of an `ActiveModel::Errors` instance.

I don't think it needs to change, just noting it here in case it becomes a common point of confusion.

File: app/models/campaign.rb
Comment (jgaskins) [inline app/models/campaign.rb:20]: :100: for sorting this

File: app/models/settings/campaign.rb
Comment (jgaskins) [inline app/models/settings/campaign.rb:7]: [nonblocking] Should we make this value editable outside of code? I don't have a great suggestion here, but the comment here seems to imply that it's something we might want to change and I'm not sure cache invalidation should require a deploy.

**Review by jgaskins — APPROVED** [review]
Just nitpicks, as usual 😄 Seriously, though, great work on this! 💯

File: app/controllers/admin/settings/campaigns_controller.rb
Comment (citizen428) [reply app/controllers/admin/settings/campaigns_controller.rb:11]: Yes. Generally, we use our custom wrapper (`ApplicationRecord#errors_as_sentence`) which does exactly what you said. These service objects — which so far I haven't refactored but kept consistent to the pre-existing one — return an array already so it's a bit different here.

File: app/models/settings/campaign.rb
Comment (citizen428) [reply app/models/settings/campaign.rb:7]: This is how the `SiteConfig` model always worked. When I wrote the generator for new settings tables I just moved this over. Once I'm done with the splitting per se I'm open to more refactorings, this one included. Deal?

**Pull request merged**
